File: src/modules/AlgoDatCourse/implemented/cpp/TaskManager.cpp
#include <iostream>
#include <queue>
#include <string>

struct Task {
    std::string name;
    int priority;

    // operator< definiert die Ordnung für std::priority_queue (Max-Heap)
    bool operator<(const Task& o) const { return priority < o.priority; }

    std::string toString() const {
        return "[" + std::to_string(priority) + "] " + name;
    }
};

int main() {
    // std::priority_queue ist standardmäßig ein MaxHeap (größtes Element oben)
    std::priority_queue<Task> pq;

    pq.push({"E-Mails beantworten", 2});
    pq.push({"Server neu starten", 10});
    pq.push({"Hauspost wegbringen", 7});
    pq.push({"Essensmarken besorgen", 1});
    pq.push({"Backup durchführen", 5});

    std::cout << "Nächste Aufgabe: " << pq.top().toString() << "\n\n";
    std::cout << "Tasks werden verarbeitet:\n";
    while (!pq.empty()) {
        std::cout << "Bearbeite: " << pq.top().toString() << "\n";
        pq.pop();
    }
}

File: src/modules/AlgoDatCourse/implemented/php/TaskManager.php
<?php

class Task
{
    public function __construct(
        public string $name,
        public int $priority
    ) {}

    public function __toString(): string
    {
        return "[{$this->priority}] {$this->name}";
    }
}

// PHP SplMaxHeap: compare() bestimmt die Ordnung.
// Gibt positiven Wert zurück, wenn $a vor $b stehen soll.
class TaskQueue extends \SplMaxHeap
{
    protected function compare(mixed $a, mixed $b): int
    {
        return $a->priority <=> $b->priority;
    }
}

$queue = new TaskQueue();
$queue->insert(new Task('E-Mails beantworten', 2));
$queue->insert(new Task('Server neu starten', 10));
$queue->insert(new Task('Hauspost wegbringen', 7));
$queue->insert(new Task('Essensmarken besorgen', 1));
$queue->insert(new Task('Backup durchführen', 5));

echo 'Nächste Aufgabe: ' . $queue->top() . "\n\n";
echo "Tasks werden verarbeitet:\n";
while (!$queue->isEmpty()) {
    echo 'Bearbeite: ' . $queue->extract() . "\n";
}
